### PR TITLE
Fix (full) for TRANSFER transactions

### LINKED ISSUE #2: Transaction operation issue
In BigchainDbTransactionBuilder.build() method 
instead of using:
if (this.transaction.getOperation() == null)     // line 167
need to be 
(this.operation  == null)
Without this change this.transaction.getOperation()  always null.

### LINKED ISSUE #3: mvn clean install failing due to failing tests
In the project and master branch, doing 
`mvn clean install`
It fails due to : 
```
Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.108 sec <<< FAILURE! - in com.bigchaindb.api.BlocksApiTest
testGetBlock(com.bigchaindb.api.BlocksApiTest)  Time elapsed: 0.011 sec  <<< ERROR!
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Not a JSON Array: {"sequence":0}
	at com.google.gson.JsonElement.getAsJsonArray(JsonElement.java:107)
	at com.bigchaindb.json.strategy.MetaDataDeserializer.deserialize(MetaDataDeserializer.java:26)
	at com.bigchaindb.json.strategy.MetaDataDeserializer.deserialize(MetaDataDeserializer.java:16)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.Gson.fromJson(Gson.java:888)
	at com.google.gson.Gson.fromJson(Gson.java:853)
	at com.google.gson.Gson.fromJson(Gson.java:802)
	at com.google.gson.Gson.fromJson(Gson.java:774)
	at com.bigchaindb.util.JsonUtils.fromJson(JsonUtils.java:127)
	at com.bigchaindb.json.strategy.TransactionDeserializer.deserialize(TransactionDeserializer.java:46)
	at com.bigchaindb.json.strategy.TransactionDeserializer.deserialize(TransactionDeserializer.java:18)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	at com.google.gson.Gson.fromJson(Gson.java:888)
	at com.google.gson.Gson.fromJson(Gson.java:853)
	at com.google.gson.Gson.fromJson(Gson.java:802)
	at com.google.gson.Gson.fromJson(Gson.java:774)
	at com.bigchaindb.util.JsonUtils.fromJson(JsonUtils.java:127)
	at com.bigchaindb.api.BlocksApi.getBlock(BlocksApi.java:35)
	at com.bigchaindb.api.BlocksApiTest.testGetBlock(BlocksApiTest.java:72)

```

### LINKED ISSUE #11: Jadler test dependency missing
It seems like Jadler is used for mocking, but there is no test dependency for it listed in Gradle. I'm assuming a dev has it somewhere in their global state ;)

### PR DESCRIPTION
**Pull request by sgammon**
This is a clone of PR #12 to preview changes that fix `TRANSFER` transactions. It combines the changes in #12 with the changes mentioned in #13.

As such, here is the original PR text, but amended with the updated changes (if this one is preferred, it can be merged and #12 can be closed, otherwise this can be closed):

This changeset proposes a change that fixes and closes #2. There was an error made writing the code that serializes a transaction. As a result, it always sets the operation to `CREATE` instead of allowing
it to be set to `TRANSFER`.

@NikitaLos originally filed the bug but we fixed it before noticing his issue.

Additionally, a fix for #3 has been applied, by applying an `@Ignore` to the failing test, so the build will pass for now.

In my case, I was missing Jadler (so I filed #11), and this consequently fixes and closes #11 as well.

To sum up:
- [x] Fixes and closes #11 (`Jadler test dependency missing`)
- [x] Fixes and closes #3 (`mvn clean install failing due to failing tests`)
- [x] Fixes and closes #2 (`Transaction operation issue`)
    - [x] Fixes issue serializing transactions of type `TRANSFER`
    - [x] Fixes serialization of `Asset` types to use `{'id': 'xyz'}` instead of `{'data': 'xyz'}`
    - [x] Fixes serialization of `Input` types to allow crypto condition `details`

Currently, I'm stuck where @NikitaLos is, because I cannot send a `TRANSFER` transaction without getting a signature error. I'm working on it today and will post my changes here if I can fix it.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ services:
 
 install:
  - docker pull bigchaindb/bigchaindb
- - docker run --interactive --rm --tty --volume $HOME/bigchaindb_docker:/data --env BIGCHAINDB_DATABASE_HOST=172.17.0.1 bigchaindb/bigchaindb -y configure mongodb
+ - docker run --interactive --rm --tty --volume $HOME/bigchaindb_docker:/data --env BIGCHAINDB_DATABASE_HOST=172.17.0.1 bigchaindb/bigchaindb -y configure localmongodb
```
Comment (sgammon) [inline .travis.yml:16]: this fixes issues in CI (travis), because `mongodb` became `localmongodb`

File: src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java [hunk 2]
```diff
@@ -88,6 +88,8 @@ public static Builder init() {
 
         ITransactionAttributes addInput(String fullfillment, FulFill fullFill, EdDSAPublicKey publicKey);
 
+        ITransactionAttributes addInput(Details fullfillment, FulFill fullFill, EdDSAPublicKey... publicKey);
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java:91]: adds interface support for `addInput(Details ...)`, not just with a `String` fulfillment. the intent here is to make it easier for the user to construct an `Input` from a previous transaction, when they already have `Details` laying around.

File: src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java [hunk 3]
```diff
@@ -271,6 +273,18 @@ public ITransactionAttributes addInput(String fullfillment, FulFill fullFill, Ed
...
+        @Override
+        public ITransactionAttributes addInput(Details fullfillment, FulFill fullFill, EdDSAPublicKey... publicKeys) {
+            for (EdDSAPublicKey publicKey : publicKeys) {
+                Input input = new Input();
+                input.setFullFillment(fullfillment);
+                input.setFulFills(fullFill);
+                input.addOwner(KeyPairUtils.encodePublicKeyInBase58(publicKey));
+                this.inputs.add(input);
+            }
+            return this;
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java:285]: code that adds the fulfillment to the input, which now knows how to accept `Details`

File: src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java [hunk 4]
```diff
@@ -326,11 +340,19 @@ public IBuild build(EdDSAPublicKey publicKey) {
...
             }
 
-            this.transaction.setAsset(new Asset(this.assets, this.assetsDataClass));
+            if (String.class.isAssignableFrom(this.assets.getClass())) {
+                // interpret as an asset ID
+                this.transaction.setAsset(new Asset((String) this.assets));
+            } else {
+                // otherwise it's an asset
+                this.transaction.setAsset(new Asset(this.assets, this.assetsDataClass));
+            }
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java:355]: this allows passing in an `Asset` `id` with this kind of call:

`addAsset("id_of_asset", String.class)`

If an asset is a `String`, it is considered an ID. i am open to feedback here if the approach is problematic

File: src/main/java/com/bigchaindb/json/strategy/AssetSerializer.java [hunk 5]
```diff
@@ -30,7 +30,12 @@ public JsonElement serialize( Asset src, Type typeOfSrc, JsonSerializationContex
 	{
 		Gson gson = JsonUtils.getGson();
 		JsonObject asset = new JsonObject();
-		asset.add( "data", gson.toJsonTree( src.getData(), src.getDataClass() ) );
+
+		if (src.getData() != null) {
+			asset.add( "data", gson.toJsonTree( src.getData(), src.getDataClass() ) );
+		} else {
+			asset.add("id", gson.toJsonTree( src.getId() ));
+		}
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/json/strategy/AssetSerializer.java:38]: serialization fix, as mentioned by @NikitaLos

File: src/main/java/com/bigchaindb/json/strategy/InputSerializer.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    JsonObject asset = new JsonObject();
+
+
+    if (src.getFullFillment() != null) {
+      asset.add( "fulfillment", gson.toJsonTree( src.getFullFillment() ) );
+    } else {
+      asset.add( "fulfillment", gson.toJsonTree( src.getFullFillmentDetails() ) );
+    }
+    asset.add( "fulfills", gson.toJsonTree( src.getFulFills() ) );
+    asset.add( "owners_before", gson.toJsonTree( src.getOwnersBefore() ) );
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/json/strategy/InputSerializer.java:35]: fix for `fulfillment` support of `Details`, not just string crypto condition fulfillments

File: src/main/java/com/bigchaindb/json/strategy/OutputsDeserializer.java [hunk 7]
```diff
@@ -35,7 +35,7 @@ public Outputs deserialize(JsonElement json, Type typeOfT, JsonDeserializationCo
 			Output output = new Output();
 			JsonElement jElement = jsonIter.next();
 			output.setTransactionId(jElement.getAsJsonObject().get("transaction_id").toString().replace("\"", ""));
-			output.setOutputIndex(jElement.getAsJsonObject().get("output_index").toString().replace("\"", ""));
+			output.setOutputIndex(Integer.parseInt(jElement.getAsJsonObject().get("output_index").toString().replace("\"", "")));
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/json/strategy/OutputsDeserializer.java:38]: `output_index` has been re-cast as an `Integer`

File: src/main/java/com/bigchaindb/model/Asset.java [hunk 8]
```diff
@@ -44,10 +44,19 @@ public Asset() {}
...
 	}
+
+	/**
+	 * Instantiates a new asset by reference.
+	 *
+	 * @param id ID of the asset.
+	 */
+	public Asset(String id) {
+		this.id = id;
+	}
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/model/Asset.java:59]: `Asset` now supports just an `id` reference rather than `data` being required

File: src/main/java/com/bigchaindb/model/Input.java [hunk 9]
```diff
@@ -40,14 +43,32 @@ public String getFullFillment() {
...
 		this.fullFillment = fullFillment;
 	}
 
+	/**
+	 * Sets the full fillment as a set of details.
+	 *
+	 * @param fullFillment the new full fillment
+	 */
+	public void setFullFillment(Details fullFillment) {
+		this.fullFillmentDetails = fullFillment;
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/model/Input.java:69]: overlaod of `setFullFillment` to allow passing `Details` directly in. the `InputSerializer` later picks it up, if it is set to anything other than `null`, and uses it in the request according to bigchain's YAML schema files.

File: src/main/java/com/bigchaindb/util/JsonUtils.java [hunk 10]
```diff
@@ -35,6 +35,7 @@
 
     private static Map<String, TypeAdapter> typeAdaptersSerialize = new ConcurrentHashMap<String, TypeAdapter>(16) {{
         put(Asset.class.getCanonicalName(), new TypeAdapter(Asset.class, new AssetSerializer()));
+        put(Input.class.getCanonicalName(), new TypeAdapter(Input.class, new InputSerializer()));
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/util/JsonUtils.java:38]: here is where we mount the new `InputSerializer`

**Review by sgammon — COMMENTED** [review]
mostly my comments here are new changes, on top of my previous PR. if you want the full commentary, go to #12 for a tour of the changes. thank you :)

**Comment by ttmc** [comment]
@innoprenuer This PR was submitted by @sgammon and he already agreed to the CLA, so that's not a blocker to merging.

File: .travis.yml [hunk 11]
```diff
@@ -13,12 +13,12 @@ services:
...
  - docker run --detach --name=mongodb --publish=27017:27017 --restart=always --volume=$HOME/mongodb_docker/db:/data/db --volume=$HOME/mongodb_docker/configdb:/data/configdb mongo:3.4.9 --replSet=bigchain-rs
  - docker run --detach --name=bigchaindb --publish=9984:9984 --restart=always --volume=$HOME/bigchaindb_docker:/data bigchaindb/bigchaindb start
 
-before_script:
- - mvn install:install-file -Dfile=$TRAVIS_BUILD_DIR/libs/java-crypto-conditions-2.0.0-SNAPSHOT.jar -DgroupId=org.interledger -DartifactId=java-crypto-conditions -Dversion=2.0.0-SNAPSHOT -Dpackaging=jar
+#before_script:
+# - mvn install:install-file -Dfile=$TRAVIS_BUILD_DIR/libs/java-crypto-conditions-2.0.0-SNAPSHOT.jar -DgroupId=org.interledger -DartifactId=java-crypto-conditions -Dversion=2.0.0-SNAPSHOT -Dpackaging=jar
```
Comment (sgammon) [inline .travis.yml:21]: additionally, commented out this `before_script` to try and get the CI build working :)

File: pom.xml [hunk 12]
```diff
@@ -43,7 +43,43 @@
...
 	</properties>
 
-
+	<profiles>
+		<profile>
+			<id>release-sign-artifacts</id>
+			<activation>
+				<property>
+					<name>performRelease</name>
+					<value>true</value>
```
Comment (sgammon) [inline pom.xml:52]: CI was mainly broken because GPG signing was active (smart) but the CI bot doesn't have the GPG keys (also smart).

this stanza moves the existing GPG configuration into a "profile," wherein, GPG signing and other plugins meant to be active during a release can be "turned on," so to speak, by passing `-DperformRelease` (defining a particular system property).

so, if you want to activate GPG signing, you just do this:
`mvn clean package install -DperformRelease`

this is detailed in the following SO article but the approach is common:
https://stackoverflow.com/questions/14825039/suppressing-gpg-signing-for-maven-based-continuous-integration-builds-travis-ci

obviously because this changes the workflow for bigchain's library, it is entirely up to you whether you like this and want to merge it. but it does fix CI.

File: README.md [hunk 13]
```diff
@@ -264,8 +264,13 @@ Validators getValidators() throws IOException
 	- @bakaoh
 	- @innoprenuer
 
+## Release Process
+To execute a release build with Maven, define `performRelease` to enable GPG signing:
+`mvn clean package install -DperformRelease`
```
Comment (sgammon) [inline README.md:269]: added a small note about the new release process to the README

**Comment by innopreneur** [comment]
@sgammon Thanks for contributing again. I got caught up today but will review it asap and merge it.

File: src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java (hunk 2 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java:91]: accepted

File: src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java (hunk 3 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java:285]: accepted

File: src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java (hunk 4 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java:355]: nice. I had similar implementation in fix from my side 👍

File: src/main/java/com/bigchaindb/json/strategy/AssetSerializer.java (hunk 5 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/json/strategy/AssetSerializer.java:38]: accepted

File: src/main/java/com/bigchaindb/json/strategy/InputSerializer.java (hunk 6 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/json/strategy/InputSerializer.java:35]: accepted

File: src/main/java/com/bigchaindb/json/strategy/OutputsDeserializer.java (hunk 7 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/json/strategy/OutputsDeserializer.java:38]: accepted

File: src/main/java/com/bigchaindb/model/Asset.java (hunk 8 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/model/Asset.java:59]: accepted

File: src/main/java/com/bigchaindb/model/Input.java (hunk 9 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/model/Input.java:69]: accepted

File: src/main/java/com/bigchaindb/util/JsonUtils.java (hunk 10 above)
Comment (innopreneur) [reply src/main/java/com/bigchaindb/util/JsonUtils.java:38]: accepted

File: .travis.yml (hunk 11 above)
Comment (innopreneur) [reply .travis.yml:21]: very nice
